### ISSUE TITLE
Implement local healthcheck endpoint

### ISSUE THREAD

**Issue by Zoullx** [favonia/cloudflare-ddns#706]
This would be to implement a local healthcheck endpoint that the Docker Service could use to check the "healthyness" of the container.

This is entirely hypothetical, but based on what I've noticed about the use of scratch and just running an executable in the container, so no `bash` or `sh`, maybe something like this:
```yaml
healthcheck:
  test: ["CMD", "ddns", "healthcheck"]
```
Basically just making it to where we could pass a parameter to the ddns executable to have it check if the service is running in the container; something along these lines.

**Comment by favonia** [comment]
@Zoullx Thanks for the suggestion. Would it be easier to check whether the Docker container is still running? For example, something like
```bash
docker container inspect -f '{{.State.Running}}' cloudflare-ddns
```

**Comment by Zoullx** [comment]
So the issue with that is that the healthcheck command for the Docker Service is ran from the context of the container. So the container would need access to the Docker CLI and the Docker Socket of the host that the container was running on for that command to work. That's why I put local healthcheck endpoint to try and differentiate it from like the healthcheck.io implementation.

**Comment by favonia** [comment]
@Zoullx Ah, I wasn't aware that Docker has a dedicated healthcheck protocol and didn't see the benefits of your suggestion. To clarify, what I suggested should definitely be run outside the Docker container.

What confuses me is the definition of "healthiness"---do failures of IP detection or IP updating (external errors) count as "unhealthy" in your usage? What are you planning to when you discover the Docker container is "unhealthy"?

**Comment by Zoullx** [comment]
So "healthyness" for a container all depends on the service. Like if there's an error detecting the IP or in setting a changed IP, if the service is still running and will try again at a later time then technically that doesn't mean that the container is "unhealthy". An "unhealthy" container essentially means that the container needs to be restarted and when the Docker Service detects that a container is "unhealthy" it will stop the old container and create a new one automatically in an attempt to end up with a "healthy" container. Essentially it orchestrates the lifetime of the container for you to ensure that the service is up and running to the best of its ability and the healthchecks built in to Docker help towards this end.

So tl;dr is that healthy means that the service that the container provides is running. Doesn't matter if it has had some errors or not, just that it's running. Unhealthy would mean that the service in the container is stopped or unable to continue to complete its task for whatever reason.

**Comment by favonia** [comment]
@Zoullx Then, I'm confident that this tool will _never_ be "unhealthy". At least I have been trying extremely hard to make sure that is the case; any deviation will be considered a serious bug. Given this, I am closing this issue. :sweat_smile: Please report any bug that might have made you believe otherwise. Still, thank you for the suggestion! :heart: I did not know that Docker has its own healthcheck protocol.

**Comment by favonia** [comment]
Actually, let me reopen this now. Is it still helpful if it always reports "healthy"?

**Comment by Zoullx** [comment]
Actually it is still helpful because timeouts can be set to where if the service doesn't respond in an allotted time, then that would be considered "unhealthy". I'm thinking something like in an environment where there are resource constraints that for whatever reason make it to where the service is running slower than normal and can't respond within a given time, then those timeouts can be adjusted through Docker to specify that the container would be considered unhealthy then.

**Comment by favonia** [comment]
@Zoullx I think I need more details about how the feature you suggested applies to this tool. My understanding is that the DDNS client is not a service (until we artificially create one just for health checking) because the way it works is to _actively_ detect IP address changes and then talks to the Cloudflare server if necessary, not passively waiting for someone else to tell it that the IP address has changed. Could you possibly help me understand the concrete scenario where you would like Docker to show that it's unhealthy. In particular, what is the "service" you were thinking about for this DDNS client? Did you encounter a case where the tool is somehow not working and you would like to manually restart it?

**Comment by favonia** [comment]
@Zoullx Just to clarify that I still want to continue the conversation to pin down the design. I am not against it, just being confused by how it should be done to be useful.

**Comment by Zoullx** [comment]
I think I have come across a perfect case for when the service could identify as being unhealthy so that the Docker service could restart it. I've ran into a couple of different situations. One where something happened with the CF_API_TOKEN environment variable that caused the process to not be able to authorize with Cloudflare and I had to manually restart the container to get the token into the container with the environment variable properly. And then another where something must have happened with the network and the process couldn't probe 1.1.1.1 or 1.0.0.1 causing it to essentially fail anyways and I had to manually restart the container to get things going through again. However, with those things said, the majority of the time I see the process recover. So I would say make it to where if the process fails to do one of these things after a certain number of times make it report as unhealthy.

**Comment by favonia** [comment]
@Zoullx Thank you for your comment. If I understand you correctly, you were talking about these scenarios:

1. **CP_API_TOKEN is invalid:** the container should end immediately, and Docker should [also give up quickly due to its restart policy.](https://docs.docker.com/config/containers/start-containers-automatically/#restart-policy-details) In this case, the container will be _stopped_ completely, which I believe is slightly better than _"running but unhealthy"._ (Or, is there a reason to keep the tool running even when it knows the updating will definitely fail?)
2. **Failing to probe 1.1.1.1 and 1.0.0.1 and then the updating failed because of that:** could you confirm that your network needs 1.0.0.1, and due to the failure to detect 1.0.0.1, the tool uses 1.1.1.1 and failed? I guess I can redesign the detection so that it will resume the testing.
3. In other cases, where the process fails to do something after a certain number of times, I do not immediately see how restarting will help. Unless you also change the configuration, the tool is designed to do exactly the same thing (modulo caching) whether you restarts it or not. That is, if something would fail, it will still fail even after restarting.

**Comment by Zoullx** [comment]
So, I think the biggest thing that you might be getting confused about, is that when the container ends up being reported as unhealthy to the Docker service, then the Docker service will automatically "recycle" that container, meaning that the Docker service will stop the unhealthy container and create a new container. This would cause it to pass the environment variables into the container again which should fix issues of the environment variables not resolving properly.

As for the failing probes, I might have to see if the service does it again and capture the log output, so I can show you what I'm talking about because I'm not sure I'll be able to explain it well enough.

**Comment by favonia** [comment]
> So, I think the biggest thing that you might be getting confused about, is that when the container ends up being reported as unhealthy to the Docker service, then the Docker service will automatically "recycle" that container, meaning that the Docker service will stop the unhealthy container and create a new container. This would cause it to pass the environment variables into the container again which should fix issues of the environment variables not resolving properly.

Hmm I'm indeed a bit confused by the line you are drawing between "unhealthy" and "stopping". The recommended setting in `README` has `restart: always`, which means that Docker will "recycle" the container as you described. Do you have `--restart always` or `restart: always` in your Docker arguments or Docker-compose configuration?

**Comment by favonia** [comment]
So, what happened in the case of invalid `CF_API_TOKEN` is the following:

1. The tool attempts to authorize it with Cloudflare and it fails.
2. The tool exits, and due to `restart: always`, Docker considers restarting it.
3. Docker might choose not to restart it if the tool exits too quickly.

The other possible configuration is `on-failure:N`, which will stop the restarting after `N` failures.

**Comment by favonia** [comment]
> As for the failing probes, I might have to see if the service does it again and capture the log output, so I can show you what I'm talking about because I'm not sure I'll be able to explain it well enough.

What will immediately help is to check whether you are using 1.1.1.1 or 1.0.0.1 _now._ If you are using 1.0.0.1 now, I can see how things could have gone wrong.

**Comment by favonia** [comment]
@Zoullx I'd like to zoom out a little bit. Is it okay to phrase your main feature request as "restarting the container when restarting might help"? If so, such situations in my opinion are bugs and should be fixed. Ideally, the tool should recover from any "non-fatal" errors; the moment it knows it's "unhealthy", it should just do what restarting could have achieved. Moreover, reporting "unhealthy" is not easier than just restarting itself, programming-wise.

I also checked the current Docker implementation, and I believe until https://github.com/moby/moby/issues/28400 is implemented, unhealthy containers will _not_ trigger the restarting. So, even if there was a scenario where reporting the "unhealthy" status is superior, the current Docker service will not restart the container for you.

**Comment by Zoullx** [comment]
> Hmm I'm indeed a bit confused by the line you are drawing between "unhealthy" and "stopping". The recommended setting in `README` has `restart: always`, which means that Docker should "recycle" the container as you described. Do you have `--restart always` or `restart: always` in your Docker arguments or Docker-compose configuration?

In terms of this, the restart always only happens if the container errors out when trying to start or stops for some unknown reason if it actually started running ever. If the container is running but the process/service/whatever you want to call it, in the container is just erroring out inside the container without a healthcheck, then the container will just keep running in that state because there's no way for the Docker service in charge of that container to know that anything is happening. In that case, the healthcheck is the magic sauce that enables that ability.

> So, what happened in the case of invalid `CF_API_TOKEN` is the following:
> 
>     1. The tool attempts to authorize it with Cloudflare and it fails.
> 
>     2. The tool exits, and due to `restart: always`, Docker considers restarting it.
> 
>     3. Docker might choose not to restart it if the tool exits too quickly.
> 
> 
> The other possible configuration is `on-failure:N`, which will stop the restarting after `N` failures.
> 
> One thing that deeply confused me is why you could restart the container to fix an invalid `CF_API_TOKEN`. You would have to change its value to make it work. Did I miss something? Are you claiming that you can fix it without changing the token?

For this, my assumption is that something happened to the value in that environment variable, probably because of environmental factors; restart of the node, power outage, what have you, but something external to Docker. Restarting the container caused it to reevaluate the environment variable. Also, I do have a restart policy set in my docker-compose files, they don't help in these cases because there's no way for the Docker service to know that anything is going on inside the container.

> What will immediately help is to check whether you are using 1.1.1.1 or 1.0.0.1 _now._ If you are using 1.0.0.1 now, I can see how things could have gone wrong.

Yeah, I mean since I've restarted the container everything has been working fine
```
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    | 🌟 Cloudflare DDNS
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    | 🥷 Dropping privileges . . .
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |    🔸 Use default PUID=1000
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |    🔸 Use default PGID=1000
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    | 📖 Reading settings . . .
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |    🔸 Use default IP4_PROVIDER=cloudflare.trace
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |    🔸 Use default UPDATE_CRON=@every 5m
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |    🔸 Use default UPDATE_ON_START=true
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |    🔸 Use default CACHE_EXPIRATION=6h0m0s
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |    🔸 Use default TTL=1
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |    🔸 Use default DETECTION_TIMEOUT=5s
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |    🔸 Use default UPDATE_TIMEOUT=30s
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    | 📖 Checking settings . . .
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    | 📖 Probing 1.1.1.1 and 1.0.0.1 . . .
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |    😊 1.1.1.1 is working. Great!
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    | 📖 Current settings:
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |    🔧 Domains and IP providers:
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |       🔸 IPv4 domains:            example.com
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |       🔸 IPv4 provider:           cloudflare.trace
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |    🔧 Scheduling:
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |       🔸 Timezone:                UTC (UTC+00 now)
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |       🔸 Update frequency:        @every 5m
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |       🔸 Update on start?         true
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |       🔸 Delete on stop?          false
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |       🔸 Cache expiration:        6h0m0s
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |    🔧 New DNS records:
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |       🔸 TTL:                     1 (auto)
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |       🔸 Proxied domains:         example.com
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |       🔸 Unproxied domains:       (none)
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |    🔧 Timeouts:
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |       🔸 IP detection:            5s
cloudflare-ddns_cloudflare-ddns.1.mq7x6sqqjj32@sdr630-58f78e80    |       🔸 Record updating:         30s
```
This is the log output, other than the 5m checks for detecting the IP, seeing that it's already set and saying that it's going to check again in 5m. Also, just to clarify, I did change my domain from what it actually is to example.com here.

> @Zoullx I'd like to zoom out a little bit. Is it okay to phrase your main feature request as "restarting the container when restarting might help"? If so, such situations in my opinion are bugs and should be fixed. Ideally, the tool should recover from any "non-fatal" errors; the moment it knows it's "unhealthy", it should just do what restarting could have achieved. Moreover, reporting the specific "unhealthy" state is not easier than just restarting itself, programming-wise.
> 
> I also checked the current Docker implementation, and I believe until [moby/moby#28400](https://github.com/moby/moby/issues/28400) is implemented, unhealthy containers will _not_ trigger the restarting. So, even if there was a scenario where reporting the "unhealthy" status is superior, the current Docker service would not restart the container for you.

Here, we basically use any status that is not considered healthy to fail the container to cause it to restart, as stated in this comment https://github.com/moby/moby/issues/28400#issuecomment-1742182004 So, while yes, you are correct in that the unhealthy status alone will not cause the container to restart because there's no specific policy for that. But that's what that issue is for and not for the fact that we can do something like `|| false` or `|| exit 1` to make the Docker service believe that the container actually failed which will in fact cause the container to be restarted, well technically recycled.

**Comment by favonia** [comment]
@Zoullx Thanks. I feel we might be talking past each other, as you seem to have repeated what you have said, and so did I. Please allow me summarize and organize what we both have said, along with my comments, to check which part might have given you the feeling that I failed to understand your point.

## Usefulness of reporting "unhealthy"

Your point is that if the tool is running but "unhealthy" under some definition, the Docker will not know it and thus could not stop it (or restart it). I understand this point. My questioning was about _when_ this will happen, and how the current design is insufficient. To clarify, the comparison was between the following design:

1. **Current design:** exit or self-recover when an error occurs, and let Docker handle restarting if it exits.
2. **Proposed design:** signal that the container is unhealthy, and let Docker handle restarting.

In my opinion, the proposed design has the drawback of using additional resources for both (1) inter-process communication for another process to report "unhealthy" and (2) keeping the container running. Moreover, it doesn't seem to be more useful than the current design.

## Difficulty of reporting "unhealthy"

Implementing a way to "detect" unhealthiness is actually tricky, and the code I could think of to signal unhealthiness will likely be more complicated than retrying things (self-recovery) or just exiting. Inter-process communication, either via system calls or file systems, is not easy in a minimum Docker image (as the released ones). This is why I'm leaning towards fixing the deviations (if any) that are not aligned with the current design. I'm happy to elaborate on what's needed to implement the health checking, but so far this point has not been challenged yet.

## Can restarting trigger re-evaluation of environmental variables?

> For this, my assumption is that something happened to the value in that environment variable, probably because of environmental factors

This really should not happen due to how Docker works, unless you went through a lot of trouble to configure your Docker that way. One main point of using Docker is to guarantee that the environment (including values of environment variables) will be very predictable. In particular, the value of `CF_API_TOKEN` will not change! The value will _always_ be the same. Thus, I have trouble seeing the usefulness of manually restarting Docker in this case.

My working theory was that there were some errors, and the process can actually recover from them. You might have chosen to restart the container, which is of course fine. However, restarting it should not be needed. It is possible that there's some hidden bug, and a detailed log will be helpful if you could recreate it.

The only case I could think of is that, the network was down when the process attempts to verify the token, and because the process exits too quickly, Docker refuses to restart it. This is filed as #722, but again, this doesn't seem to apply in your case; you seem to be claiming that the value of some environment variable might have changed, but that should never happen with Docker.

## Can failure to probe 1.1.1.1 

> the process couldn't probe 1.1.1.1 or 1.0.0.1 causing it to essentially fail 

Your most recent logging shows that 1.1.1.1 is working is for you, so maybe in one of your runs the probing failed due to the networking being temporarily down. In the case of such a failure, the process will stick to 1.1.1.1 when the probing failed. Therefore, as the previous case, my working theory is that you restart the container when it can just self-recover.

The only case I could think of was that, maybe the probing fails and the process sticks to 1.1.1.1, and yet your network needs the 1.0.0.1 workaround to work. This is filed as #721 but it does not apply in your case.

## General points about the reporting

I coded the process in a way that it should self-recover from _any_ temporary error, including power outage, networking being down, Cloudflare servers being down, DNS zones being reorganized, etc. It is designed to be resilient enough that none of these errors can stop it. It will just try again next time (maybe after 5 minutes). I took this resilience very seriously and treat any deviations as bugs. I am very interested in checking the two cases where you claimed manual restarting is needed, which to me means the process was buggy, but the analysis you provided does not seem to match how Docker works or what the current code is. If you still have the detailed logging, that will be tremendously useful. Thanks.

**Comment by favonia** [comment]
@Zoullx Hi, I haven't heard from you for a while and I still don't have an actionable item.

What I gathered is that you have encountered some problem (perhaps some bug) in the past, and you believe a local healthcheck endpoint would help. However, based on the information you provided so far, I strongly believe the hypothetical healthcheck (no matter how it could be implemented) will not help you. I am worried that we might have the [XY communication problem](https://en.wikipedia.org/wiki/XY_problem) here---that is, you might have encountered some real issue but a healthcheck point is _not_ the solution. If that is the case, our discussion might not be very helpful.

I thus wonder if you could provide more details about the errors you encountered or the difficulties you have experienced in the past that have caused you to restart the updater. Only with those details I can see what would be the best solution to your problem and/or assess whether there's a real bug. I feel we have been wasting too much time on debating whether a hypothetical situation is possible, etc., and perhaps the more important question is to ask what was really causing your trouble and whether the solution can really solve your problem (not some hypothetical situation).

Once you are ready to restart the conversation, please kindly re-open this issue with the details of the errors or difficulties and then we can start from there. For now, because there are no actionable items, I would have to close this issue.

**Comment by favonia** [comment]
## Retrospective maintainer notes

### What drove the original closure

This issue was closed as **not planned** based on the reasoning at that time:

- I try not to add features unless they are materially useful for this project.
- For this DDNS updater, a basic container healthcheck mostly duplicates liveness (`running`) and does not clearly improve update reliability.
- We already support outbound monitoring integrations (for example, Healthchecks and Uptime Kuma), which are better signals for “did updates actually happen?”.

### New understanding (added later)

This is a later reflection, not part of the original closure rationale:

- In reality, there are two primary ways to use healthchecks in this context:
  - In some frameworks (e.g., Docker Compose), a service can wait until another service is healthy via `depends_on` with `service_healthy`. This is _startup gating._
  - In some frameworks (e.g., Kubernetes), a container can be restarted when liveness/startup checks keep failing.
- _Startup gating_ is low value here, since no other services should depend on this updater in most cases. (Its main use is to update IP addresses, not to create DNS records. `DELETE_ON_STOP=1` can remove all addresses when exiting, which may then trigger record creation on the next start.)
- For this DDNS tool, common failures are external (network/provider/API), so restarting usually does not fix the root cause. For that reason, “external update failure” should generally not be treated as “container unhealthy.”

### Practical takeaway

- Practically, marking a container unhealthy is most useful when startup gating or restarting helps; failures due to transient external reasons should usually not indicate a container is unhealthy.
- `docker container inspect -f '{{.State.Running}}' <container>` is fine for liveness.
- For progress/success visibility, logging (for local checking) and outbound monitoring (for remote checking) remain the most meaningful approach for this project.

**Comment by fiveangle** [comment]
The notion that awaiting startup of the container that updates a known-incorrect DDNS entry as "low value" is arguable. In my case, the choices are:

1. Add some arbitrary delay and accept the longer-than-necessary startup delay of the downstream container that publishes its DDNS host to the candidate balancer.
2. Add no delay, and craft some double-checking of the endpoint from the load balancer side to double-confirm it is in fact functioning at the DDNS hostname. (Which it won't be, so add another back-off-delay algorithm, i.e., custom hand-jamming workarounds that are always prone to failure due to human error/oversight since it is buried in a remote shell half a world away).
3. Add an additional 2 containers (Uptime Kuma and perhaps an MQTT endpoint) to serve as a similar Rube Goldberg coordination service, adding more failure-prone plumbing (although at least less obscure).
4. Just accept transient failures for users during the startup of the new node. (Wait, it's not 1992... nope, that just won't do).
5. Craft my own custom image for cloudflare-ddns (everyone's favorite answer on how to mung turn-key microservices back into locally maintained installations).

While all are possible options, I just wanted to seed the idea that the purpose of the `health_check` of any container managed by Compose does not live in a static, arbitrary definition handed down by the almighty Docker spec. It is really about what is "_healthy_" to that container, its maintainer, and its grateful users.

K.I.S.S. is indeed a notion lost on most these days, so I earnestly applaud your stance there. But the option of a "phone call" method for active DDNS update microservice status rather than a "mailbox" method seems like an obvious killer feature for a component whose sole purpose is to try and minimize DNS-to-DynamicIP lookup times. By keeping the DDNS client "simple" and not supporting this basic Compose function, it forces many users into crafting entirely more complex mechanisms (along with all the potential pitfalls that come with one-off fragile coordination plumbing) just to get native Docker Compose functionality that would "just work" with this relatively modest feature.

Ideally, users could choose from two status endpoints for the container to test: a "the daemon is up and attempting update... NOW" endpoint (the bare minimum, and granted, relatively low-value, but in the context of a VPS that is hyper memory and IO-constrained, it could still mean the difference between minutes vs. seconds of getting a resource online). But the golden ticket "last host update was successfully completed" endpoint check would dispense with all the tomfoolery workarounds and "just work" for the bulk of users in need of it. That flexibility would be handsomely appreciated.

Thanks for coming to my T.E.D. Talk : )

But seriously, thanks for keeping this working these many years, and being a reasonable gatekeeper of simplicity for the countless folks relying on your project. I just hope my TL;DR exposé punctuates just how valuable of a position you are in to help simplify life for them even more. If you'd like me to draft a first-pass at the feature, I wouldn't mind, although I suspect it would grate you more than help, especially for something you'd be shackled with maintaining. When your beard starts to show more gray than not, the "just hack it all together to make it work" approach doesn't have the same appeal as it does to a greenhorn. 

Cheers, and thanks again,

-=dave

**Comment by favonia** [comment]
@fiveangle Thank you for explaining your use case and for your great encouragement. The original request and most of the discussion here focused on restarting the container after failures. Your use case---waiting for a DNS update to complete before starting a dependent service---raises different design questions. This would be worth discussing in a separate issue. In the meantime, here are my initial thoughts:

---

Am I understanding correctly that you want **at least one successful update cycle** since startup, with all configured domains and WAF lists updated successfully?

Previously, I considered startup gating low value because I was picturing a long-running setup where the IP address can change during operation. In that setting, the load balancer already needs to handle temporary mismatches between the current IP and the DNS records. Waiting for the initial update would not address those later mismatches, so I assumed the same handling would cover startup as well. If the IP instead stays fixed for the node's lifetime, one-time mode (`UPDATE_CRON=@once`) might be a simpler fit: complete the update before starting the dependent service.

Could you open a new issue briefly describing your use case? Feel free to link to your comment here rather than repeat the details.
